### PR TITLE
#166151257 Integrate with Fitbit

### PR DESCRIPTION
**Pull request by willeswa**
#### What does this PR do?
It integrates Fitbit API
#### Description of Task to be completed?
- [x] Configures a FitBit application
- [x] Implement OAuth2 for the implementation
#### How should this be manually tested?
- [x] Create a Fitbit account: set the redirect uri, get the client_id and secret and set them in your env
- [x] Clone the repo and set it up for development.
- [x] Checkout out to the `ft-fitbit-support-166151257` branch & create an account
- [x] On your Wger dashboard, you should be able to see your name on the upper right corner
- [x] Click on the dropdown and click on Sync Fitbit
- [x] You should be able to authorize and be redirected back to the dashboard
#### Any background context:
When creating a testing application on fitbit, the call backurl should point to this endpoint ```/en/user/sync_fitbit/```
#### Story ID
[166151257](https://www.pivotaltracker.com/story/show/166151257)
#### Screenshots
Screen Shot 2019-06-13 at 22 52 12
Screen Shot 2019-06-13 at 22 52 24
Screen Shot 2019-06-13 at 22 52 41
Screen Shot 2019-06-13 at 22 53 09
Screen Shot 2019-06-13 at 23 07 27
Screen Shot 2019-06-13 at 23 23 24

### CODE REVIEW

File: wger/core/views/user.py [hunk 1]
```diff
@@ -40,6 +43,11 @@
 from django.conf import settings
 from rest_framework.authtoken.models import Token
 
+from fitbit import FitbitOauth2Client
+import base64
+import requests
```
Comment (wangonya) [inline wger/core/views/user.py]: You can move the import statements here to the top.

Comment (ceexon) [reply wger/core/views/user.py]: Why should he make the change yet flake 8 has not raised any complaints.

Comment (wangonya) [reply wger/core/views/user.py]: This is just to conform to pep8 standards. According to [this](https://pep8.org/#imports) resource, import statements should be grouped as follows:

1. standard library imports
2. related third party imports
3. local application/library specific imports

He has them mixed up here.

**Review by wangonya — COMMENTED** [review]
Your PR description message seems to be a bit different. Any reason you chose to go with this format?
Also, the PR heading does not conform to standards. Kindly refer to https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices

**Comment by willeswa** [comment]
> Your PR description message seems to be a bit different. Any reason you chose to go with this format?
> Also, the PR heading does not conform to standards. Kindly refer to https://github.com/andela/bestpractices/wiki/Git-naming-conventions-and-best-practices

The PR format follows this convention introduced during the Deferral program. Refer to https://docs.google.com/document/d/1k1vITs3zDuVS7l2gCY5ML-kqAOq7euQIBlWrNonwhYw/edit?usp=sharing. I just realized, the above were team conventions. I have updated the PR.

**Review by jmusila — COMMENTED** [review]
I found the PR description a little bit hard to follow when testing locally, maybe you can add more descriptions or if there are environment variables to be exported, you can add them in the `.env_sample` file, that is, if there are more variables.

File: wger/core/views/user.py [hunk 2]
```diff
@@ -323,6 +330,82 @@ def preferences(request):
         return render(request, 'user/preferences.html', template_data)
 
 
+@login_required
+def sync_fitbit(request):
+    """Authorizes the connection with Wger
+
+    """
+
+    if settings.WGER_SETTINGS['FITBIT_CLIENT_ID'] is None:
```
Comment (wekesa931) [inline wger/core/views/user.py:339]: Impressive work! 🙌🏼 You could consider testing for the availability of a secret key too.

File: wger/core/views/user.py [hunk 3]
```diff
@@ -323,6 +330,82 @@ def preferences(request):
...
+        response = requests.post(
+            fitbit_auth_client.request_token_url, form, headers=headers).json()
+
+        if "access_token" in response:
+            token = response['access_token']
+            headers['Authorization'] = 'Bearer ' + token
+            date = (datetime.date.today()).strftime('%Y-%m-%d')
+
+            response = requests.get(
+                'https://api.fitbit.com/1/user/-/body/log/weight/date/' + date + '.json',
```
Comment (wekesa931) [inline wger/core/views/user.py:376]: Is this uri hardcoded or is it customized for every user that logs into wger?

File: wger/core/views/user.py (hunk 2 above)
Comment (willeswa) [reply wger/core/views/user.py:339]: > Impressive work! 🙌🏼 You could consider testing for the availability of a secret key too.

Working on it

**Comment by willeswa** [comment]
> I found the PR description a little bit hard to follow when testing locally, maybe you can add more descriptions or if there are environment variables to be exported, you can add them in the `.env_sample` file, that is, if there are more variables.

Thank you for pointing out this. I have updated the .sample.env. Does it make it easier now?

File: wger/core/views/user.py (hunk 3 above)
Comment (willeswa) [reply wger/core/views/user.py:376]: > Is this uri hardcoded or is it customized for every user that logs into wger?

It varies for each user. I think a lot has been abstracted by Fitbit. It seems that once you authorize Wger to get Fitbit data, that uri will somehow be custom to your account only.

File: wger/core/views/user.py [hunk 4]
```diff
@@ -323,6 +330,82 @@ def preferences(request):
         return render(request, 'user/preferences.html', template_data)
 
 
+@login_required
+def sync_fitbit(request):
+    """Authorizes the connection with Wger
+
```
Comment (wangonya) [inline wger/core/views/user.py:336]: The formatting here is a bit off. You can either bring the lower triple quotes up to line 335 or take the `Authorizes the connection with Wger` down to line 336.

Comment (willeswa) [reply wger/core/views/user.py:336]: > The formatting here is a bit off. You can either bring the lower triple quotes up to line 335 or take the `Authorizes the connection with Wger` down to line 336.

Noted.

**Review by ceexon — APPROVED** [review]

File: .env_sample [hunk 5]
```diff
@@ -9,4 +9,7 @@ export SECRET_KEY=''
 
 
 export RECAPTCHA_PUBLIC_KEY=''
-export RECAPTCHA_PRIVATE_KEY=''
\ No newline at end of file
+export RECAPTCHA_PRIVATE_KEY=''
+export FITBIT_CLIENT_ID=22DRQ3
+export FITBIT_CLIENT_SECRET=a9c286f26f3228fc572a53b93bea5b73
+export ROOT_URL='http://127.0.0.1:8000'
```
Comment (caveinn) [inline .env_sample]: It is good practice to have sensitive information such as the keys not visible on the codebase. I would advise having this on the manually configurable env settings on both travis and heroku.

**Review by caveinn — COMMENTED** [review]
Awesome work. Just check on the comment I left.

File: .env_sample (hunk 5 above)
Comment (willeswa) [reply .env_sample]: This is a sample .env and the secret codes are not real or valid. They are supposed to show what is to be added to the real .env

**Review by caveinn — APPROVED** [review]
LGTM
